File: clients/Ruby-client/lib/solidb/geo_client.rb
module SoliDB
  class GeoClient
    def initialize(client)
      @client = client
    end

    def create_index(collection, name:, fields:, geo_json: nil)
      args = { database: @client.database, collection: collection, name: name, fields: fields }
      args[:geo_json] = geo_json unless geo_json.nil?
      @client.send_command("create_geo_index", args)
    end

    def list_indexes(collection)
      @client.send_command("list_geo_indexes", database: @client.database, collection: collection) || []
    end

    def delete_index(collection, index_name)
      @client.send_command("delete_geo_index", database: @client.database, collection: collection, index_name: index_name)
      nil
    end

    def near(collection, latitude:, longitude:, radius:, limit: nil)
      args = {
        database: @client.database,
        collection: collection,
        latitude: latitude,
        longitude: longitude,
        radius: radius
      }
      args[:limit] = limit if limit
      @client.send_command("geo_near", args) || []
    end

    def within(collection, geometry:)
      @client.send_command("geo_within", database: @client.database, collection: collection, geometry: geometry) || []
    end

    def distance(lat1:, lon1:, lat2:, lon2:)
      @client.send_command("geo_distance", lat1: lat1, lon1: lon1, lat2: lat2, lon2: lon2)
    end

    def intersects(collection, geometry:)
      @client.send_command("geo_intersects", database: @client.database, collection: collection, geometry: geometry) || []
    end
  end
end

File: clients/PHP-client/src/GeoClient.php
<?php

namespace SoliDB;

class GeoClient
{
    private Client $client;

    public function __construct(Client $client)
    {
        $this->client = $client;
    }

    public function createIndex(string $collection, string $name, array $fields, ?bool $geoJson = null): array
    {
        $args = [
            'database' => $this->client->getDatabase(),
            'collection' => $collection,
            'name' => $name,
            'fields' => $fields
        ];
        if ($geoJson !== null) {
            $args['geo_json'] = $geoJson;
        }
        $res = $this->client->sendCommand('create_geo_index', $args);
        return $res['data'] ?? [];
    }

    public function listIndexes(string $collection): array
    {
        $res = $this->client->sendCommand('list_geo_indexes', [
            'database' => $this->client->getDatabase(),
            'collection' => $collection
        ]);
        return $res['data'] ?? [];
    }

    public function deleteIndex(string $collection, string $indexName): void
    {
        $this->client->sendCommand('delete_geo_index', [
            'database' => $this->client->getDatabase(),
            'collection' => $collection,
            'index_name' => $indexName
        ]);
    }

    public function near(string $collection, float $latitude, float $longitude, float $radius, ?int $limit = null): array
    {
        $args = [
            'database' => $this->client->getDatabase(),
            'collection' => $collection,
            'latitude' => $latitude,
            'longitude' => $longitude,
            'radius' => $radius
        ];
        if ($limit !== null) {
            $args['limit'] = $limit;
        }
        $res = $this->client->sendCommand('geo_near', $args);
        return $res['data'] ?? [];
    }

    public function within(string $collection, array $geometry): array
    {
        $res = $this->client->sendCommand('geo_within', [
            'database' => $this->client->getDatabase(),
            'collection' => $collection,
            'geometry' => $geometry
        ]);
        return $res['data'] ?? [];
    }

    public function distance(float $lat1, float $lon1, float $lat2, float $lon2): float
    {
        $res = $this->client->sendCommand('geo_distance', [
            'lat1' => $lat1,
            'lon1' => $lon1,
            'lat2' => $lat2,
            'lon2' => $lon2
        ]);
        return $res['data'] ?? 0.0;
    }

    public function intersects(string $collection, array $geometry): array
    {
        $res = $this->client->sendCommand('geo_intersects', [
            'database' => $this->client->getDatabase(),
            'collection' => $collection,
            'geometry' => $geometry
        ]);
        return $res['data'] ?? [];
    }
}
